File: Distribution/Samples/Async/ExcelTaskUtil.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using ExcelDna.Integration;

namespace ExcelDna.AsyncSample
{
    internal static class ExcelTaskUtil
    {
        // Careful - this might only work as long as the task is not shared between calls, since cancellation cancels that task
        // Another implementation via Reactive Extension is Task.ToObservable() (in System.Reactive.Linq.dll) with RxExcel
        public static IExcelObservable ToExcelObservable<TResult>(this Task<TResult> task)
        {
            if (task == null)
            {
                throw new ArgumentNullException("task");
            }

            return new ExcelTaskObservable<TResult>(task);
        }

        public static IExcelObservable ToExcelObservable<TResult>(this Task<TResult> task, CancellationTokenSource cts)
        {
            if (task == null)
            {
                throw new ArgumentNullException("task");
            }

            return new ExcelTaskObservable<TResult>(task, cts);
        }

        public static object RunTask<TResult>(string callerFunctionName, object callerParameters, Func<CancellationToken, Task<TResult>> taskSource)
        {
            return ExcelAsyncUtil.Observe(callerFunctionName, callerParameters, delegate
                {
                    var cts = new CancellationTokenSource();
                    var task = taskSource(cts.Token);
                    return new ExcelTaskObservable<TResult>(task, cts);
                });
        }

        public static object RunTask<TResult>(string callerFunctionName, object callerParameters, Func<Task<TResult>> taskSource)
        {
            return ExcelAsyncUtil.Observe(callerFunctionName, callerParameters, delegate
            {
                var task = taskSource();
                return new ExcelTaskObservable<TResult>(task);
            });
        }

	    public static object RunAsTask<TResult>(string callerFunctionName, object callerParameters, Func<CancellationToken, TResult> function)
	    {
            return RunTask(callerFunctionName, callerParameters, cancellationToken => Task.Factory.StartNew<TResult>(() => function(cancellationToken), cancellationToken));
	    }

	    public static object RunAsTask<TResult>(string callerFunctionName, object callerParameters, Func<TResult> function)
	    {
		    return RunTask(callerFunctionName, callerParameters, () => Task.Factory.StartNew<TResult>(function));
	    }

        // Helper class to wrap a Task in an Observable - allowing one Subscriber.
        class ExcelTaskObservable<TResult> : IExcelObservable
        {
            readonly Task<TResult> _task;
            readonly CancellationTokenSource _cts;

            public ExcelTaskObservable(Task<TResult> task)
            {
                _task = task;
            }

            public ExcelTaskObservable(Task<TResult> task, CancellationTokenSource cts)
                : this(task)
            {
                _cts = cts;
            }

            public IDisposable Subscribe(IExcelObserver observer)
            {
                switch (_task.Status)
                {
                    case TaskStatus.RanToCompletion:
                        observer.OnNext(_task.Result);
                        observer.OnCompleted();
                        break;
                    case TaskStatus.Faulted:
                        observer.OnError(_task.Exception.InnerException);
                        break;
                    case TaskStatus.Canceled:
                        observer.OnError(new TaskCanceledException(_task));
                        break;
                    default:
                        _task.ContinueWith(t =>
                        {
                            switch (t.Status)
                            {
                                case TaskStatus.RanToCompletion:
                                    observer.OnNext(t.Result);
                                    observer.OnCompleted();
                                    break;
                                case TaskStatus.Faulted:
                                    observer.OnError(t.Exception.InnerException);
                                    break;
                                case TaskStatus.Canceled:
                                    observer.OnError(new TaskCanceledException(t));
                                    break;
                            }
                        });
                        break;
                }

                // Check for cancellation support
                if (_cts != null)
                {
                    return new CancellationDisposable(_cts);
                }
                // No cancellation
                return DefaultDisposable.Instance;
            }
        }

        sealed class DefaultDisposable : IDisposable
        {

            public static readonly DefaultDisposable Instance = new DefaultDisposable();
            
            // Prevent external instantiation
            DefaultDisposable()
            {
            }

            public void Dispose()
            {
                // no op
            }
        }

        sealed class CancellationDisposable : IDisposable
        {

            readonly CancellationTokenSource cts;
            public CancellationDisposable(CancellationTokenSource cts)
            {
                if (cts == null)
                {
                    throw new ArgumentNullException("cts");
                }

                this.cts = cts;
            }

            public CancellationDisposable()
                : this(new CancellationTokenSource())
            {
            }

            public CancellationToken Token
            {
                get { return cts.Token; }
            }

            public void Dispose()
            {
                cts.Cancel();
            }
        }
    }
}

File: Distribution/Samples/Async/ExcelTaskUtil.vb
Imports System.Runtime.CompilerServices
Imports System.Threading
Imports System.Threading.Tasks
Imports ExcelDna.Integration

Public Module ExcelTaskUtil
    ' Another implementation via Reactive Extension is Task.ToObservable() (in System.Reactive.Linq.dll) with RxExcel
    <Extension()> _
    Public Function ToExcelObservable(Of TResult)(task As Task(Of TResult)) As IExcelObservable
        If task Is Nothing Then
            Throw New ArgumentNullException("task")
        End If

        Return New ExcelTaskObservable(Of TResult)(task)
    End Function

    <Extension()> _
    Public Function ToExcelObservable(Of TResult)(task As Task(Of TResult), cts As CancellationTokenSource) As IExcelObservable
        If task Is Nothing Then
            Throw New ArgumentNullException("task")
        End If

        Return New ExcelTaskObservable(Of TResult)(task, cts)
    End Function

    Public Function RunTask(Of TResult)(callerFunctionName As String, callerParameters As Object, taskSource As Func(Of CancellationToken, Task(Of TResult))) As Object
        Return ExcelAsyncUtil.Observe(callerFunctionName, callerParameters, Function()
                                                                                Dim cts = New CancellationTokenSource()
                                                                                Dim task = taskSource(cts.Token)
                                                                                Return New ExcelTaskObservable(Of TResult)(task, cts)
                                                                            End Function)
    End Function

    Public Function RunTask(Of TResult)(callerFunctionName As String, callerParameters As Object, taskSource As Func(Of Task(Of TResult))) As Object
        Return ExcelAsyncUtil.Observe(callerFunctionName, callerParameters, Function()
                                                                                Dim task = taskSource()
                                                                                Return New ExcelTaskObservable(Of TResult)(task)
                                                                            End Function)
    End Function

    Public Function RunAsTask(Of TResult)(callerFunctionName As String, callerParameters As Object, func As Func(Of CancellationToken, TResult)) As Object
        Return RunTask(callerFunctionName, callerParameters,
                       Function(cancellationToken) Task.Factory.StartNew(Of TResult)(Function() func(cancellationToken), cancellationToken))
    End Function

    Public Function RunAsTask(Of TResult)(callerFunctionName As String, callerParameters As Object, func As Func(Of TResult)) As Object
        Return RunTask(callerFunctionName, callerParameters, Function() Task.Factory.StartNew(Of TResult)(func))
    End Function

    ' Helper class to wrap a Task in an Observable - allowing one Subscriber.
    Private Class ExcelTaskObservable(Of TResult)
        Implements IExcelObservable
        ReadOnly _task As Task(Of TResult)
        ReadOnly _cts As CancellationTokenSource

        Public Sub New(task As Task(Of TResult))
            _task = task
        End Sub

        Public Sub New(task As Task(Of TResult), cts As CancellationTokenSource)
            Me.New(task)
            _cts = cts
        End Sub

        Public Function Subscribe(observer As IExcelObserver) As IDisposable Implements IExcelObservable.Subscribe
            Select Case _task.Status
                Case TaskStatus.RanToCompletion
                    observer.OnNext(_task.Result)
                    observer.OnCompleted()
                    Exit Select
                Case TaskStatus.Faulted
                    observer.OnError(_task.Exception.InnerException)
                    Exit Select
                Case TaskStatus.Canceled
                    observer.OnError(New TaskCanceledException(_task))
                    Exit Select
                Case Else
                    _task.ContinueWith(Sub(t)
                                           Select Case t.Status
                                               Case TaskStatus.RanToCompletion
                                                   observer.OnNext(t.Result)
                                                   observer.OnCompleted()
                                                   Exit Select
                                               Case TaskStatus.Faulted
                                                   observer.OnError(t.Exception.InnerException)
                                                   Exit Select
                                               Case TaskStatus.Canceled
                                                   observer.OnError(New TaskCanceledException(t))
                                                   Exit Select
                                           End Select
                                       End Sub)
                    Exit Select
            End Select

            ' Check for cancellation support
            If _cts IsNot Nothing Then
                Return New CancellationDisposable(_cts)
            End If
            ' No cancellation
            Return DefaultDisposable.Instance
        End Function
    End Class

    Private NotInheritable Class DefaultDisposable
        Implements IDisposable

        Public Shared ReadOnly Instance As New DefaultDisposable()

        ' Prevent external instantiation
        Private Sub New()
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            ' no op
        End Sub
    End Class

    Private NotInheritable Class CancellationDisposable
        Implements IDisposable

        ReadOnly cts As CancellationTokenSource

        Public Sub New(cts As CancellationTokenSource)
            If cts Is Nothing Then
                Throw New ArgumentNullException("cts")
            End If

            Me.cts = cts
        End Sub

        Public Sub New()
            Me.New(New CancellationTokenSource())
        End Sub

        Public ReadOnly Property Token As CancellationToken
            Get
                Return cts.Token
            End Get
        End Property

        Public Sub Dispose() Implements IDisposable.Dispose
            cts.Cancel()
        End Sub
    End Class

End Module